### PR TITLE
Big refactor

### PR DESCRIPTION
**Pull request by boygirl**
cc/ @exogen 

This is finally clean and stable enough for a PR. I think this version is a lot more flexible, but that means it can also get broken more easily. There are probably a lot more places that need warnings.

### CODE REVIEW

File: src/components/victory-chart.jsx [hunk 1]
```diff
@@ -881,157 +583,89 @@ export default class VictoryChart extends React.Component {
...
-        domain={this.domain[axis]}
-        range={this.range[axis]}
-        orientation={this.props.axisOrientation[axis]}
-        tickCount={this.props.tickCount[axis]}
-        tickValues={this.tickValues[axis]}
-        tickFormat={this.tickFormat[axis]}
-        style={style}/>
+      <svg style={this.style.parent}>
+        {this.getNewChildren()}
+      </svg>
```
Comment (exogen) [inline src/components/victory-chart.jsx]: Nice and small and clean now, but a feature was removed – seems like we might want to support rendering many charts into a single `<svg>`? But doesn't seem like a huge deal to me.

File: src/util.js [hunk 2]
```diff
@@ -12,5 +12,15 @@ module.exports = {
     return _.every(collection, function (item) {
       return _.isString(item);
     });
+  },
+
+  isArrayOfArrays: function (collection) {
+    return _.isArray(collection) && _.every(collection, function (item) {
+      return _.isArray(item);
+    });
+  },
```
Comment (exogen) [inline src/util.js]: I wonder if there's a way we can re-use React's `propTypes` validators for stuff like this? Do they expose any validation functions? So we could use `React.PropTypes.arrayOf(React.PropTypes.array)` somehow.

Comment (boygirl) [inline src/util.js]: yeah, totally, but the prop this is needed for takes either an array or an array of arrays :/

File: src/components/victory-chart.jsx [hunk 3]
```diff
@@ -881,157 +583,89 @@ export default class VictoryChart extends React.Component {
...
+    if (_.isFunction(y)) {
+      return _.map(xArray, (x) => {
+        return {x, y: y(x)};
+      });
+    } else {
+      const n = _.min([xArray.length, y.length]);
+      return _.map(_.take(xArray, n), (x, index) => {
+        return { x, y: y[index]};
+      });
+    }
```
Comment (exogen) [inline src/components/victory-chart.jsx]: I find all these small functions and nested lodash calls hard to follow. Can we break some of these out into `util.js` so this functionality has a name and clear parameters?

File: src/util.js (hunk 2 above)
Comment (exogen) [inline src/util.js]: Oh for sure I get that part – I mean instead of the contents of this function being this lodash `isArray` and `every` methods, maybe React has a function we could use so it would look more like:

```
React.doesValidate(collection, React.PropTypes.arrayOf(React.PropTypes.array));
```

...which just seems cleaner and more declarative to me.

File: src/components/victory-chart.jsx [hunk 4]
```diff
@@ -498,371 +337,234 @@ export default class VictoryChart extends React.Component {
...
+      // TODO: clean up cumulative max / min check assumptions.
+      const cumulativeMax = _.reduce(formattedData, (memo, dataset) => {
+        const localMax = (_.max(_.pluck(dataset, "y")));
+        return localMax > 0 ? memo + localMax : memo;
+      }, 0);
+      const cumulativeMin = _.reduce(formattedData, (memo, dataset) => {
+        const localMin = (_.min(_.pluck(dataset, "y")));
+        return localMin < 0 ? memo + localMin : memo;
+      }, 0);
+      return [_.min([min, cumulativeMin]), _.max([max, cumulativeMax])];
```
Comment (exogen) [inline src/components/victory-chart.jsx]: Seems like this could be a util function too.

File: src/components/victory-chart.jsx [hunk 5]
```diff
@@ -329,167 +94,241 @@ export default class VictoryChart extends React.Component {
...
+    });
+  }
+
+  getAxisComponents(axis) {
+    return _.filter(this.childComponents, (child) => {
+      return axis === "x" ?
+      child.props.independentAxis && child.type.displayName === "VictoryAxis" :
+      child.type.displayName === "VictoryAxis" && !child.props.independentAxis;
+    });
+  }
```
Comment (exogen) [inline src/components/victory-chart.jsx]: Hm, I don't like how much this whole set of methods depends on `displayName` being something specific. What if someone subclasses `VictoryAxis` and wants to use that? It just seems too fragile and specific to me.

A few options that come to mind:

```javascript
child instanceof VictoryAxis
```

or:

```javascript
class VictoryAxis extends React.Component {
  isAxis: true
  ...
}
```

or

```javascript
class VictoryAxis extends React.Component {
  role: "axis"
  ...
}
```

The latter is probably my favorite. Then every other Victory Component can get a `role` too and we'll be able to determine the "structure" of the chart (like we're doing in these methods) no matter which components are used (subclassed/customized/or not).
